### PR TITLE
K8SPSMDB-299: Smart update shard

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-299](https://jira.percona.com/browse/K8SPSMDB-299)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| storage | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| upgrade | passed |
| one-pod | passed |
| init-deploy | passed |
| self-healing | passed |
| monitoring | passed |
| upgrade-consistency | failed |
| limits | passed |
| scaling | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| upgrade | passed |
| upgrade-consistency | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| demand-backup | passed |
| scheduled-backup | passed |
| users | failed |

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

File: pkg/psmdb/mongo/mongo.go [hunk 1]
```diff
@@ -123,12 +114,63 @@ func RSStatus(ctx context.Context, client *mongo.Client) (Status, error) {
...
+	return SwitchBalancer(ctx, client, false)
+}
+
+func SwitchBalancer(ctx context.Context, client *mongo.Client, state bool) error {
+	res := OKResponse{}
+
+	command := "balancerStop"
+	if state {
+		command = "balancerStart"
+	}
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/psmdb/mongo/mongo.go]: https://github.com/percona/percona-server-mongodb-operator/pull/449/commits/3bcaf6cf7817ff35002403108e6edbb23400470a

`IsBalancerRunning` and `switchBalancer` have different result type

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| one-pod | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | failed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-449-fd7f7e62

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

File: pkg/controller/perconaservermongodb/smart.go [hunk 2]
```diff
@@ -116,11 +140,151 @@ func (r *ReconcilePerconaServerMongoDB) smartUpdate(cr *api.PerconaServerMongoDB
...
+			Namespace: cr.Namespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				"app.kubernetes.io/name":       "percona-server-mongodb",
+				"app.kubernetes.io/instance":   cr.Name,
+				"app.kubernetes.io/managed-by": "percona-server-mongodb-operator",
+				"app.kubernetes.io/part-of":    "percona-server-mongodb",
+				"app.kubernetes.io/component":  "mongos",
+			}),
+		},
+	)
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/smart.go:229]: can we check Deployment here? and check state identically to StatefulSet?
```
if s.Status.UpdatedReplicas < s.Status.Replicas {
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/smart.go (hunk 2 above)
Comment (blez) [reply pkg/controller/perconaservermongodb/smart.go:246]: https://github.com/percona/percona-server-mongodb-operator/pull/449/commits/c9397a3f456df78e2d9b7ef67e8b5caf6212a2c8

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | failed |
| init-deploy | passed |
| upgrade | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | failed |
| self-healing | passed |
| security-context | failed |
| monitoring | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| scheduled-backup | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

**Review by delgod — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | failed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| one-pod | failed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | failed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-449-6b2834d2

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by qJkee — CHANGES_REQUESTED** [review]
We should also check the `sh.isBalancerRunning()` command.
Because when it return `true` it transfers chunks between shard and we need to wait until it return false

**Comment by blez** [comment]
@qJkee , no we should not
https://docs.mongodb.com/manual/reference/command/balancerStop/#dbcmd.balancerStop
```
If a balancing round is in progress, the operation waits for balancing to complete.
```

**Comment by qJkee** [comment]
will the `StopBalancer` function wait for process completion then?

**Comment by blez** [comment]
yes, as documentation tells

**Review by delgod — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Review by dAdAbird — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| scheduled-backup | failed |

**Pull request merged**